Add Editor Analytics for package.

vickycl opened the pull request:
## Purpose of this PR

**Ticket/Jira #:** [USDU-384](https://jira.unity3d.com/browse/USDU-384)

<!-- Description of feature/change. Links to screenshots, design docs, user docs, etc. Remember reviewers may be outside your team, and not know your feature/area that should be explained more. -->

Links: [Spike document](https://docs.google.com/document/d/1hl0kTdK-9Gqkkbm1ywfAHAHT7z_OcYvBmPgvUayDW4A/edit#) | [Editor Data Collection form](https://docs.google.com/spreadsheets/d/1bOhBgplhthCxSfIIJI2dTxp6uGW1SOYVNbqk6S_55mw/edit#gid=1274740786)

Begins adding basic analytics to understand package usage. Includes:
- Initialization of the plugin
- Import events + relevant context:
    - success
    - file format
    - time taken
    - features used
- Export events + relevant context:
    - success
    - file format
    - is it only overrides
    - time taken
- Recorder export

This PR doesn't capture _everything_ we would like to, as called out in the [doc](https://docs.google.com/document/d/1hl0kTdK-9Gqkkbm1ywfAHAHT7z_OcYvBmPgvUayDW4A/edit#heading=h.a5hv0kmbhb0). It's mostly an MVP to build upon. Payloads and Variants events standout as features that might be relatively simple but useful to collect.


## Testing

**Functional Testing status:**

<!-- Explanation of what's tested, how tested and existing or new automation tests. Can include manual testing by self and/or QA. Specify test plans. Rarely acceptable to have no testing.  -->

Currently attempting to write automated tests. Manually tested that the basic workflows collate the data correctly using the Analytics Debugger, and viewed the data on BigQuery, all looks ok.

**Performance Testing status:**

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->

None yet.

## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:** medium
<!-- (Minimal / Low / Medium / High) -->

**Halo Effect:** low
<!-- (Minimal / Low / Medium / High) -->

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [x] Add entry in CHANGELOG.md _(if applicable)_

vickycl committed:
Initial ideas for Editor instrumentation.

- Adds a UsdEditorAnalytics class to control reporting events to Editor Analytics
- Includes a few changes to import and export so that we can report a status, but I'm not sure this is particularly useful. I'm leaving it in just in case.

vickycl committed:
Fixes for compile errors.

Commit:
Apply formatting changes

vickycl committed:
Include meta file

vickycl committed:
Improvements to analytics:

- Adds timings to import and export events
- Adds analytics for recorder export events
- changes primMap so that instanceRoots are accessible

I moved the SendEvent calls into the SceneImporter/Exporter files so that they capture re-imports and API calls. This does mean it will be harder/ messier to report whether the calls came from button presses or users using the APIs, as well as whether they are importing to GameObjects or Prefabs.

Known issues:
- point instancer not correctly reported
- not sure how to report issues/ failures exporting from recorder
- on failures in some cases we no longer know the file extension
- EditorAnalytics should be stripped out of runtime
- need some code cleanup

Commit:
Apply formatting changes

vickycl committed:
Code clean up.

vickycl committed:
Add todo note.

Also // TODO:
- Manual testing

vickycl committed:
Small fixes for builds.

- Compile out body of APIs in non-editor
- Fix for build time clashes with plugin

vickycl committed:
Update the Analytics preprocessor directive

Relevant docs: https://docs.unity3d.com/ScriptReference/Analytics.Analytics.html

vickycl committed:
Fix for PointInstances and potential CI fix.

vickycl committed:
Remove all 'upmNoDefaultPackages'

This stops the UnityAnalytics package being added, which leads to script compilation errors. I'm not sure exactly why we (and the CI templates) pass this option, I'll have to investigate.

michaeljblain commented on package/com.unity.formats.usd/Editor/Scripts/UsdBuildPostProcess.cs:
```diff
@@ -45,20 +46,28 @@ public static void OnPostprocessBuild(BuildTarget target, string pathToBuiltProj
+            try
             {
-                Directory.CreateDirectory(destination);
+                if (!Directory.Exists(destination))
```
Can tidy this up a bit to only have 1 call to Directory.CreateDirectory.

vickycl replied:
Oh yeah, whoops!

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/InstanceImporter.cs:
```diff
@@ -163,6 +163,8 @@ public static void BuildSceneInstances(PrimMap primMap, SceneImportOptions optio
                 primMap.AddInstance(goInstance);
             }
+
+            primMap.ContainsPointInstances = instanceCount > 0;
```
Could we move this up a level and just set it once if we have >0 point instancer prims (rather than calling it for every point instancer). My assumption was that we'd just capture whether point instancer prims exist within the imported usd  and not check the point instancer data itself.

vickycl replied:
Done :)

michaeljblain reviewed (commented):
Looks like a solid first pass. Some minor comments. Just wondering how the known issues can be addressed.

vickycl committed:
Addressed PR feedback.

vickycl committed:
Fix for recorder bug with null context.

vickycl committed:
Revert BuildPostProcess changes to see if it is causing the Standalone CI issue.

vickycl committed:
Added a failure case for Recorder, removed comments.

vickycl committed:
Update CHANGELOG.md

SimonBoorer commented on package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs:
```diff
@@ -0,0 +1,199 @@
+# if ENABLE_CLOUD_SERVICES_ANALYTICS && UNITY_EDITOR
```
Copyright notice should probably be at the beginning of the file

SimonBoorer commented on package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs:
```diff
@@ -0,0 +1,199 @@
+            if (sUsdEditorAnalyticsEvents[eventName])
+                return true;
+
+            AnalyticsResult result = EditorAnalytics.RegisterEventWithLimit(eventName, k_MaxEventsPerHour, k_MaxNumberOfElements, k_VendorKey);
```
What are the other possible results for AnalyticsResult? Should we log if registration fails?

vickycl replied:
The other results are the 'reasons' it can fail, eg UnsupportedPlatform. We could debug log it for the user, but I'm not sure whether they could necessarily do anything about it, and as the analytics has failed to initialize then we can't report it back to ourselves. 

Maybe it's worth Debug.Logging anyway, so that we could get slightly more informative bug reports if a user decided to report it.

vickycl replied:
I've added logging for users on both the registering and sending of events :)

SimonBoorer commented on package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs:
```diff
@@ -0,0 +1,199 @@
+
+        public static void SendImportEvent(string fileExtension, float timeTakenS, ImportResult result)
+        {
+            Debug.Log($"Send event called");
```
Remove me?

CraigHutchinson commented on package/com.unity.formats.usd/Runtime/InitUsd.cs:
```diff
@@ -56,9 +60,13 @@ public static bool Initialize()
             {
                 Debug.LogException(ex);
+                analyticsTimer.Stop();
+                UsdEditorAnalytics.SendUsageEvent(false, analyticsTimer.ElapsedMilliseconds * 0.001f);
```
Minor: Could use API `analyticsTimer.Elapsed.TotalSeconds` here and elsewhere

vickycl replied:
Ah, I missed that API :)

SimonBoorer commented on package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs:
```diff
@@ -0,0 +1,199 @@
+            public string FileExtension;
+            public bool ExportSucceeded;
+            public bool OnlyOverrides;
+            public int FrameCount;
```
Should we be recording time taken here too or is that not valid in this context?

vickycl replied:
Hmm, I guess it would be valid, while exporting via the recorder we do some work every frame so we could keep a cumulative total of the time taken to export to USD over all of the frames.

SimonBoorer commented on package/com.unity.formats.usd/Editor/Scripts/Behaviors/UsdMenu.cs:
```diff
@@ -14,9 +14,9 @@
-using UnityEngine;
 using UnityEditor;
 using USD.NET;
+using Debug = UnityEngine.Debug;
```
Not used?

vickycl replied:
It's used in a function that I haven't changed, where previous it just used the UnityEngine include instead. I'll revert this change :)

SimonBoorer commented on package/com.unity.formats.usd/Runtime/InitUsd.cs:
```diff
@@ -12,10 +12,11 @@
-using UnityEngine;
 using USD.NET;
 using USD.NET.Unity;
+using Debug = UnityEngine.Debug;
```
Not used?

vickycl replied:
EDIT: this is because adding System.Diagnostics added ambiguity for Debug, so in this case just saves changing all the existing Debug.Log() calls to UnityEngine.Debug.Log().

SimonBoorer commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/ImportHelpers.cs:
```diff
@@ -8,7 +9,9 @@
 using UnityEngine;
 using USD.NET;
 using USD.NET.Unity;
+using Debug = UnityEngine.Debug;
```
Not used?

vickycl replied:
In this case I've swapped the `using System.Diagnostics;` for `using Stopwatch = System.Diagnostics.Stopwatch;` for consistency and removed the `using Debug = UnityEngine.Debug;` - is that better?

SimonBoorer commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs:
```diff
@@ -14,11 +14,13 @@
 using USD.NET;
 using USD.NET.Unity;
 using Unity.Jobs;
+using Debug = UnityEngine.Debug;
```
Not used?

SimonBoorer commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/PrimMap.cs:
```diff
@@ -48,19 +48,23 @@ public struct InstanceRoot
         // Objects at the root of an instanced sub-tree.
         // Instances may be found in masters as well.
-        private Dictionary<SdfPath, InstanceRoot> m_instanceRoots = new Dictionary<SdfPath, InstanceRoot>();
+        public Dictionary<SdfPath, InstanceRoot> InstanceRoots = new Dictionary<SdfPath, InstanceRoot>();
```
Seems like this change is no longer necessary with the addition of ContainsPointInstances?

vickycl replied:
Yep- forgot to revert it. Thanks :)

SimonBoorer commented on package/com.unity.formats.usd/Runtime/Scripts/Behaviors/UsdAsset.cs:
```diff
@@ -590,14 +590,22 @@ public void ExportOverrides(Scene sceneInWhichToStoreTransforms)
 
+            bool success = false;
+
+            System.Diagnostics.Stopwatch analyticsTimer = new System.Diagnostics.Stopwatch();
```
Would be good to standardise on either using System.Diagnostics or System.Diagnostics.Stopwatch

same with Unity.Formats.USD.UsdEditorAnalytics.ImportResult

vickycl replied:
What do you mean by this? Do you mean specifically in the 'using's sometimes I've done it and sometimes I haven't?

SimonBoorer replied:
Yeah pretty much, change looks good thanks!

SimonBoorer reviewed (commented):
General approach looks good, some minor nitpicks

CraigHutchinson commented on package/com.unity.formats.usd/Dependencies/USD.NET/collections/SampleCollection.cs:
```diff
@@ -29,6 +29,8 @@ public class SampleCollection<T> : IEnumerable<SampleEnumerator<T>.SampleHolder>
         private SdfPath[] m_paths;
         private Scene m_scene;
 
+        public int GetLength() => m_paths.Length;
```
Minor: Length is a property and could be defined as such: `public int Length { get =>  m_paths.Length; }`

vickycl replied:
My reasoning for calling it 'GetLength' was so that it was explicit that it was retrieving something from a member, and not to be confused with an Array.Length. I might be overthinking this though. Do you think making it a property would be nicer?

CraigHutchinson replied:
Its more just a C# convention as the 'get' (and 'set') prefix is handled as part of the C# language using properties. 

Really boils down to your personal preference and the .Net library itself isn't consistent but once you set the name its harder to change. Some points for official reference
https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/bzwdh01d(v=vs.71)?redirectedfrom=MSDN#properties-vs-methods

vickycl replied:
Yep, seems like in this case Length is actually the way to go. Thanks :)

CraigHutchinson commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/PrimMap.cs:
```diff
@@ -48,19 +48,23 @@ public struct InstanceRoot
         // Objects at the root of an instanced sub-tree.
         // Instances may be found in masters as well.
-        private Dictionary<SdfPath, InstanceRoot> m_instanceRoots = new Dictionary<SdfPath, InstanceRoot>();
+        public Dictionary<SdfPath, InstanceRoot> InstanceRoots = new Dictionary<SdfPath, InstanceRoot>();
```
Comment: Not sure if this is intended to be public, was private before change and coudln't spot usage outside of this.

vickycl replied:
You're right- I forgot to revert this when I switched to the ContainsPointInstances flag instead :)

SimonBoorer commented:
Another thought - I think you can use timer.Elapsed.Seconds rather than timer.ElapsedMilliseconds * 0.001f - but might it make more sense to capture time in milliseconds anyway?

CraigHutchinson reviewed (approved):
No major observations, only  minor changes that may be addressed

vickycl committed:
Address PR feedback.

vickycl committed:
Small fixes

vickycl committed:
Addressed feedback- renaming, adding cumulative time taken to recorder analytics

vickycl committed:
Fix hardcoded extension to be the same as others

vickycl committed:
Added note to documentation.

vickycl committed:
Fix for animation streaming and other small things.

- Added a flag to stop analytics being sent for every frame of animation replay
- renamed the events to remove the redundant "u"
- changed the time takens to all be in ms instead of seconds

vickycl committed:
Code cleanup.

Commit:
Apply formatting changes

vickycl committed:
Merge branch 'dev' into add-editor-analytics

vickycl committed:
Add an enum for import type

This allows us to combine functionality to ignore 'streaming' imports, and include whether the import is a reimport in a more extensible way.

michaeljblain commented on package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs:
```diff
@@ -22,6 +22,14 @@
+    public enum ImportType
+    {
+        Initial,
+        Reimport,
```
To avoid confusion with `bool IsReImport` in the Analytics class perhaps we call this something like "Replace" or "ForceRebuild" or similar.

michaeljblain reviewed (commented):
Looks good. Given the work is already done we could actually store output the import type as part of the analytics (instead of a bool).

vickycl committed:
Updates to analytics

- adds a dedicated event for reimports
- changes timing to float to allow sub-1ms timings

michaeljblain commented on package/com.unity.formats.usd/Runtime/InitUsd.cs:
```diff
@@ -61,12 +61,12 @@ public static bool Initialize()
                 Debug.LogException(ex);
                 analyticsTimer.Stop();
-                UsdEditorAnalytics.SendUsageEvent(false, analyticsTimer.ElapsedMilliseconds);
+                UsdEditorAnalytics.SendUsageEvent(false, (float)analyticsTimer.Elapsed.TotalMilliseconds);
```
Is there a reason why we cast this to float instead of just sending through the original double value?

michaeljblain reviewed (approved):
Looks good to me. Just one question about float vs double.

vickycl committed:
Improved error handling for Import Analytics

michaeljblain reviewed (changes requested):
This seems to now correctly indicate import failure for our test case. Longer term we should address error handling in a more thorough way. 
Some minor feedback: 
* the commit message doesn't indicate the purpose of the inclusion of the `SceneTemplateSettings.json` file.
* I think we still should store the ms as a double instead of casting to float.

vickycl committed:
Change stopwatch times to double and removed accidentally committed file.

vickycl commented:
@michaeljblain 

> This seems to now correctly indicate import failure for our test case. Longer term we should address error handling in a more thorough way. Some minor feedback:
> 
> * the commit message doesn't indicate the purpose of the inclusion of the `SceneTemplateSettings.json` file.

This was an accident :) thanks for pointing it out!

> * I think we still should store the ms as a double instead of casting to float.

Done

michaeljblain reviewed (approved):
Thanks for the changes.

vickycl committed:
Merge branch 'dev' into add-editor-analytics

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -114,3 +114,8 @@ We do not currently support the following:
+
+## Analytics
+
+We collect high level analytics of the features used in the package, to help us understand adoption and prioritize future development work. You can see exactly what data points are collected in `Common\UsdEditorAnalytics.cs`.
```
Comment: Maybe something for future, but maybe there is a contact email for analytics at Unity so as/if corporate entities have concerns over our collections they can make queries to address the concern that could arise from analytics being present. i.e. Opt-in or out GDPR sort of stuff

vickycl replied:
These are just standard analytics the same as any other in the Unity Editor, as described in the official documentation: https://docs.unity3d.com/Manual/EditorAnalytics.html

CraigHutchinson commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/ExportHelpers.cs:
```diff
@@ -65,7 +66,7 @@ public static Scene InitForSave(string filePath)
 
             analyticsTimer.Stop();
-            UsdEditorAnalytics.SendExportEvent(Path.GetExtension(scene.FilePath), analyticsTimer.ElapsedMilliseconds * 0.001f, success);
+            UsdEditorAnalytics.SendExportEvent(Path.GetExtension(scene.FilePath), (float)analyticsTimer.Elapsed.TotalMilliseconds, success);
```
Moderate: Unnecessary float cast

vickycl replied:
Good spot :)

CraigHutchinson commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs:
```diff
@@ -278,14 +277,19 @@ static SceneImporter()
             if (scene == null)
             {
-                UsdEditorAnalytics.SendImportEvent("", .0f, importResult);
+                UsdEditorAnalytics.SendImportEvent("", 0.0f, importResult);
```
Trivial: Could use 0 or 0.0 as in other places as this is now 'double' type

CraigHutchinson reviewed (commented):
Looks good but a really minor (float) cast left over from earlier revision

vickycl committed:
Small PR fixes
